modules/biobank/php/specimencontroller.class.inc

```diff
@@ -0,0 +1,531 @@
+     */
+    private $db, $user, $dao;
+
+    /**
+     * Specimen Controller Class constructor
+     *
+     * @param Database $db   Database Instance.
+     * @param User     $user User Instance.
+     */
+    function __construct(\Database $db, \User $user)
+    {
+        $this->db   = $db;
+        $this->user = $user;
+        $this->dao  = $this->getDataProvisioner();
+    }
+
+    /**
+     * Get all Specimen Objects permitted by the Specimen Data Provisioner (DAO).
+     *
+     * @return array $specimens All permissable Specimen Objects.
+     */
+    public function getInstances() : array
+    {
+        $this->validatePermission('GET');
+        $specimens  = [];
+        $specimenIt = $this->dao->execute($this->user);
+        foreach ($specimenIt as $id => $specimen) {
+            $specimens[$id] = $specimen;
+        }
+
+        return $specimens;
+    }
+
+    /**
+     * This method takes in a list of specimen and container arrays from the
+     * front-end. The specimens are validated while the containers are
+     * sent to their respective controller to be validated and saved. The ids of
+     * the saved containers are set to the specimen instances which are then
+     * sent to the DAO to be saved.
+     *
+     * @param array $list a list of specimen and container arrays to be created.
+     *
+     * @return void
+     */
+    public function createInstances(array $list) : void
+    {
+        $this->validatePermission('POST');
+
+        foreach ($list as $i => $item) {
+            $containerList[$i] = $item['container'];
+            $specimenList[$i]  = $this->prepare($item['specimen']);
+        }
+
+        $contController = new ContainerController($this->db, $this->user);
+        $containerIds   = $contController->createInstances($containerList);
+
+        foreach ($specimenlist as $i => $specimen) {
+            $specimen->setContainerId($containerIds[$i]);
+            $this->dao->saveInstance($specimen);
+        }
+    }
+
+    // TODO: Decide if this function should take a list of specimens, rather
+    // than a singular specimen.
+    /**
+     * This method takes in a specimen array from the front-end, validates it,
+     * instantiates it and then sends it to the DAO to be saved to the
+     * database.
+     *
+     * @param array $specimenArray specimen array to be created
+     *
+     * @return void
+     */
+    public function updateInstance(array $specimenArray) : void
+    {
+        $this->validatePermission('PUT');
+        $specimen = $this->prepare($specimenArray);
+        $this->dao->saveInstance($specimen);
+    }
+
+    /**
+     * Converts data into a specimen intance validates the specimen.
+     *
+     * @param array $specimenArray Specimen data to be instantiated
+     *
+     * @return Specimen $specimen to be submitted
+     */
+    private function prepare(array $specimenArray) : Specimen
+    {
+        $specimen = $this->dao->getInstanceFromArray($specimenArray);
```

Comment:
I mean, it knows the server-side field names which in this case happen to be the same as the client-side field names. If you wanted we could have another function in the controller that is called `toServerSideArrayFromClientSideArray()` but it wouldn't really do anything.

Reply:
the way I see it there are 3 different "languages" 
`(1) front end field names -> (2) Transfer object variables -> (3) SQL fieldnames`

Given that the DAO defined here is an SQL specific it should know (and should be the only one that knows) what the SQL field names are. Also, since the DAO knows (like other classes) how to manipulate the TO then it should know the TO variable associations to the SQL field names. However, I don't see a reason for it to know anything about the names on the frontend. If we later revamp the front end or go API only the DAO should not be affected at all it sits one layer below that.

The DAO should do mappings `2 -> 3` and `3 -> 2` for sure, but `getInstanceFromArray()` is mapping `1 -> 2` and that I feel should be either the controller or the TO itself doing it (preferably the controller this way the TO is also independent from the front-end being altered)